ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/ModelConformanceValidator.scala

```diff
@@ -24,121 +19,133 @@ import com.daml.ledger.participant.state.kvutils.{Conversions, Err}
+
+        for {
+          stepResult <- consumeValidationResult(
+            validationResult,
+            transactionEntry,
+            commitContext,
+            rejections,
+          )
+          finalStepResult <- validateCausalMonotonicity(stepResult, commitContext, rejections)
+        } yield finalStepResult
+      })
+  }
 
-        try {
-          engine
-            .validate(
-              transactionEntry.submitters.toSet,
-              SubmittedTransaction(transactionEntry.transaction),
-              transactionEntry.ledgerEffectiveTime,
-              commitContext.participantId,
-              transactionEntry.submissionTime,
-              transactionEntry.submissionSeed,
-            )
-            .consume(
-              lookupContract(transactionEntry, commitContext),
-              lookupPackage(commitContext),
-              lookupKey(commitContext, knownKeys),
-            )
-            .fold(
-              err =>
-                rejections.buildRejectionStep(
-                  transactionEntry,
-                  rejectionReasonForValidationError(err),
-                  commitContext.recordTime,
-                ),
-              _ => StepContinue[DamlTransactionEntrySummary](transactionEntry),
-            )
-        } catch {
-          case err: Err.MissingInputState =>
-            logger.warn(
-              "Model conformance validation failed due to a missing input state (most likely due to invalid state on the participant)."
-            )
+  private def consumeValidationResult(
+      validationResult: Result[Unit],
+      transactionEntry: DamlTransactionEntrySummary,
+      commitContext: CommitContext,
+      rejections: Rejections,
+  )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
+    try {
+      val stepResult = for {
+        contractKeyInputs <- transactionEntry.transaction.contractKeyInputs.left
+          .map(rejectionForKeyInputError(transactionEntry, commitContext.recordTime, rejections))
+        _ <- validationResult
+          .consume(
+            lookupContract(commitContext),
+            lookupPackage(commitContext),
+            lookupKey(contractKeyInputs),
+          )
+          .left
+          .map(error =>
             rejections.buildRejectionStep(
               transactionEntry,
-              RejectionReasonV0.Disputed(err.getMessage),
+              RejectionReasonV0.Disputed(error.msg),
               commitContext.recordTime,
             )
-        }
-      })
-  }
-
-  private def contractIsActive(
-      transactionEntry: DamlTransactionEntrySummary,
-      contractState: DamlContractState,
-  ): Boolean = {
-    val activeAt = Option(contractState.getActiveAt).map(parseTimestamp)
-    !contractState.hasArchivedAt && activeAt.exists(transactionEntry.ledgerEffectiveTime >= _)
+          )
+      } yield ()
+      stepResult.fold(identity, _ => StepContinue(transactionEntry))
+    } catch {
+      case err: Err =>
```

Comment:
Here we need to differentiate between different "consumption" errors:
* `Err.MissingInputState` => `Inconsistent`
* `Err.DecodeError` => `Disputed`

Reply:
Done